Add nullptr checks before accessing children in getUnspecialized

augusto2112 opened the pull request:
The demangler already has an error mechanism to report if demangling failed. Add null pointer checks before every access in Demangle::getUnspecialized, and return an error if the child doesn't exist.

rdar://110141007

augusto2112 commented:
@swift-ci smoke test

adrian-prantl reviewed (approved):
This is great, thanks!

al45tair reviewed (commented):
It would be better to use `DEMANGLER_ASSERT()` here, I think, since these nodes are supposed to have children and not having means there's a bug somewhere else in the demangling code.

`DEMANGLER_ASSERT()` returns a `ManglingError` in the runtime and in non-asserts builds, but in asserts builds it'll raise an assertion failure, which arguably we'd like to do in this case I think.

e.g. instead of
```
if (!node->hasChildren())
  return MANGLING_ERROR(ManglingError::BadNominalTypeKind, node);
```
we could do
```
DEMANGLER_ASSERT(node->hasChildren(), node);
```

augusto2112 committed:
Add nullptr checks before accessing children in getUnspecialized

The demangler already has an error mechanism to report if demangling
failed. Add null pointer checks before every access in
Demangle::getUnspecialized, and return an error if the child doesn't
exist.

rdar://110141007

augusto2112 commented:
@swift-ci smoke test
